exonum/src/storage/indexes_metadata.rs

```diff
@@ -0,0 +1,252 @@
+    Entry,
+    KeySet,
+    List,
+    SparseList,
+    Map,
+    ProofList,
+    ProofMap,
+    ValueSet,
+}
+
+impl From<IndexType> for u8 {
+    fn from(tt: IndexType) -> Self {
+        use self::IndexType::*;
+        match tt {
+            Entry => 0,
+            KeySet => 1,
+            List => 2,
+            SparseList => 3,
+            Map => 4,
+            ProofList => 5,
+            ProofMap => 6,
+            ValueSet => 7,
+        }
+    }
+}
+
+impl From<u8> for IndexType {
+    fn from(num: u8) -> Self {
+        use self::IndexType::*;
+        match num {
+            0 => Entry,
+            1 => KeySet,
+            2 => List,
+            3 => SparseList,
+            4 => Map,
+            5 => ProofList,
+            6 => ProofMap,
+            7 => ValueSet,
+            invalid => {
+                panic!(
+                    "Unreachable pattern ({:?}) while constructing table type. \
+                    Storage data is probably corrupted",
+                    invalid
+                )
+            }
+        }
+    }
+}
+
+impl CryptoHash for IndexType {
+    fn hash(&self) -> Hash {
+        u8::from(*self).hash()
+    }
+}
+
+impl StorageValue for IndexType {
+    fn into_bytes(self) -> Vec<u8> {
+        u8::from(self).into_bytes()
+    }
+
+    fn from_bytes(value: Cow<[u8]>) -> Self {
+        u8::from_bytes(value).into()
+    }
+}
+
+impl<'a> Field<'a> for IndexType {
+    unsafe fn read(buffer: &'a [u8], from: Offset, to: Offset) -> Self {
+        u8::read(buffer, from, to).into()
+    }
+
+    fn write(&self, buffer: &mut Vec<u8>, from: Offset, to: Offset) {
+        u8::from(*self).write(buffer, from, to)
+    }
+
+    fn field_size() -> Offset {
+        u8::field_size()
+    }
+
+    fn check(
+        buffer: &'a [u8],
+        from: CheckedOffset,
+        to: CheckedOffset,
+        latest_segment: CheckedOffset,
+    ) -> ::std::result::Result<CheckedOffset, EncodingError> {
+        u8::check(buffer, from, to, latest_segment)
+    }
+}
+
+impl json::ExonumJson for IndexType {
+    fn deserialize_field<B: WriteBufferWrapper>(
+        value: &json::reexport::Value,
+        buffer: &mut B,
+        from: Offset,
+        to: Offset,
+    ) -> Result<(), Box<Error>>
+    where
+        Self: Sized,
+    {
+        let v = value.as_u64().ok_or("Can't cast json as u64")? as u8;
+        buffer.write(from, to, v);
+        Ok(())
+    }
+
+    fn serialize_field(&self) -> Result<json::reexport::Value, Box<Error + Send + Sync>> {
+        Ok(json::reexport::Value::from(u8::from(*self)))
+    }
+}
+
+pub fn assert_read(name: &str, index_type: IndexType, view: &Snapshot) {
+    let metadata = BaseIndex::indexes_metadata(view);
+    if let Some(value) = metadata.get::<_, IndexMetadata>(name) {
+        let stored_type = value.index_type();
+        assert_index_type(name, index_type, stored_type);
+    }
+}
+
+pub fn assert_write(name: &str, index_type: IndexType, view: &mut Fork) {
```

Comment:
It panics if conditions are wrong, as for me, it can be called "assertion". Additionally, we already have some similar functions (`assert_valid_name` for example).